content-security-policy/support/testharness-helper.sub.js

```diff
@@ -0,0 +1,47 @@
+const Host = {
+  SAME_ORIGIN: "same-origin",
+  CROSS_ORIGIN: "cross-origin",
+};
+
+const PolicyHeader = {
+  CSP: "echo-policy.py?policy=",
+  EMBEDDING_CSP: "echo-embedding-csp.py",
+} 
+
+function generateURL(host, path) {
+  var url = new URL("http://{{host}}:{{ports[http][0]}}/content-security-policy/support/");
+  url.hostname = host == Host.SAME_ORIGIN ? "{{host}}" : "{{domains[天気の良い日]}}";
+  url.pathname += path;
+
+  return url.toString();
+}
+
+function generateRedirect(host, target) {
+  var url = "http://{{host}}:{{ports[http][0]}}/common/redirect.py?location=" + encodeURIComponent(target);
+  if (host == Host.CROSS_ORIGIN) 
+    url = "http://{{domains[天気の良い日]}}:{{ports[http][0]}}/common/redirect.py?location=" + encodeURIComponent(target);
+
+  return url.toString();
+}
+
+function assert_embedding_csp(t, url, csp, expected, newCsp=null, newSrc=null) {
+  var i = document.createElement('iframe');
+  if (csp) 
+    i.csp = csp;
+  i.src = url;
+
+  window.addEventListener('message', t.step_func(e => {
+    if (e.source != i.contentWindow || !('embedding_csp' in e.data))
+        return;
+    if (newSrc && newCsp)
+      expected = newCsp;
+    assert_equals(expected, e.data['embedding_csp']);
+    t.done();
+  }));
+
+  document.body.appendChild(i);
+  if (newCsp)
+    i.csp = newCsp;
+  if (newSrc)
+    i.src = newSrc;
```

Comment:
> I was thinking it would be sufficient to check that the latest CSP is always used since we append the iframe and then change it.

Right, but you're setting up a race condition by setting `i.src` to `url`, then appending the element (which kicks off a load), then modifying `i.src` to `newSrc` (which kicks off another load, probably before the first load has finished, but maybe not). That means that the `message` event you're listening for might come from either the first or second load, based purely upon how the browser decided to execute your code.

If you want to verify that setting the `csp` attribute doesn't affect an already-loaded page, I think a better way of doing so would be to wait for the `load` event. If you want to verify that setting the `csp` attribute before a page completes its load, then I think you need to set the expectations more strictly (e.g. drop the `e.data['expected_csp'] == null` carveout).